# Project 2: Data Profiling and Preparation

## 1. Conduct Data Access and Profiling:
- To begin, read the survey dataset into a pandas DataFrame.
- Then, create a subset of the DataFrame to include only passenger satisfaction data about the Business Class. Name the created DataFrame subset “df_Business”.
- Collect information about the “df_Business” dataset and its validity. Report the following:
  1. The number of observations (or rows) and the number of variables (or columns) in the dataset.
  2. The name and data type of each data column.
  3. The unique values of each column.
  4. The number of missing values of each column.
  5. Summary statistics for each data column.
  6. Frequency distribution (list the number of times each unique value appears) for each data column.
  7. The number of fully duplicated data rows.
- Based on the information that you collected, provide a list of the issues that the data contains as an annotation.

### To begin, read the survey dataset into a pandas DataFrame.

In [1]:
#Importing the needed packages
import pandas as pd
import numpy as np

In [2]:
#Importing the survey dataset
df = pd.read_csv ('Passengers_Satisfaction_Survey (1).csv')
df.head()

Gender      Customer Type   Age   Type of Travel     Class  \
0    Male     Loyal Customer  69.0  Personal Travel  Business   
1    Male  Disloyal Customer  20.0  Business travel       Eco   
2    Male  Disloyal Customer  41.0  Business travel  Business   
3    Male     Loyal Customer  52.0  Business travel  Business   
4  Female  Disloyal Customer  21.0  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0          -5421.0                    3.0                                2.0   
1              0.0                    3.0                                3.0   
2              0.0                    1.0                                1.0   
3              0.0                    2.0                                4.0   
4              0.0                    4.0                                0.0   

  Ease of Online booking Food and drink  Inflight entertainment  \
0                      3              5                     1.0   
1                      3              2                    10.0   
2                    NaN            NaN                     NaN   
3                      4              1                     2.0   
4                    NaN            NaN                     NaN   

  Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                1          3.0                        14.0   
1                4          2.0                         0.0   
2              NaN          NaN                         NaN   
3                2          4.0                         0.0   
4              NaN          NaN                         NaN   

   Arrival Delay in Minutes  Satisfaction  
0                       0.0  Dissatisfied  
1                       0.0  dissatisfied  
2                       NaN       neutral  
3                       6.0  Dissatisfied  
4                       NaN       neutral

In [3]:
df.shape

(104910, 16)

### Then, create a subset of the DataFrame to include only passenger satisfaction data about the Business Class. Name the created DataFrame subset “df_Business”.

In [4]:
Class_counts = df['Class'].value_counts()
Class_counts
#There is no format error about the word "Business", so we can directly filter business class by the name of 'Business'. 

Business    50000
Eco         47313
Eco Plus     7597
Name: Class, dtype: int64

In [5]:
df_Business = df[(df["Class"] == "Business")] 
df_Business.reset_index(drop=True, inplace=True)
df_Business.head()

Gender      Customer Type   Age   Type of Travel     Class  \
0    Male     Loyal Customer  69.0  Personal Travel  Business   
1    Male  Disloyal Customer  41.0  Business travel  Business   
2    Male     Loyal Customer  52.0  Business travel  Business   
3  Female  Disloyal Customer  21.0  Business travel  Business   
4  Female     Loyal Customer  57.0  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0          -5421.0                    3.0                                2.0   
1              0.0                    1.0                                1.0   
2              0.0                    2.0                                4.0   
3              0.0                    4.0                                0.0   
4              0.0                    3.0                                4.0   

  Ease of Online booking Food and drink  Inflight entertainment  \
0                      3              5                     1.0   
1                    NaN            NaN                     NaN   
2                      4              1                     2.0   
3                    NaN            NaN                     NaN   
4                      4              5                     2.0   

  Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                1          3.0                        14.0   
1              NaN          NaN                         NaN   
2                2          4.0                         0.0   
3              NaN          NaN                         NaN   
4                3          2.0                         0.0   

   Arrival Delay in Minutes  Satisfaction  
0                       0.0  Dissatisfied  
1                       NaN       neutral  
2                       6.0  Dissatisfied  
3                       NaN       neutral  
4                       0.0  Dissatisfied

### Collect information about the “df_Business” dataset and its validity. Report the following:


#### 1. The number of observations (or rows) and the number of variables (or columns) in the dataset.

In [6]:
df_Business.shape
#We can observe that there are 50,000 rows and 16 columns in our dataset.

(50000, 16)

#### 2. The name and data type of each data column.

In [7]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  obj

#### 3. The unique values of each column.

In [8]:
cols = df_Business.columns
# for each column
for col in cols:
    print(col)
    # get a list of unique values
    unique = df_Business[col].unique()

    # if number of unique values is less than 30, print the values. Otherwise print the number of unique values
    if len(unique)<30:
        print(unique, '\n====================================\n\n')
    else:
        print(str(len(unique)) + ' unique values', '\n====================================\n\n')

Gender
['Male' 'Female' 'F' 'M' nan] 


Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 


Age
75 unique values 


Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 


Class
['Business'] 


Flight Distance
3574 unique values 


Inflight wifi service
[ 3.  1.  2.  4.  0.  5. nan] 


Departure/Arrival time convenient
[ 2.  1.  4.  0.  5.  3. nan] 


Ease of Online booking
['3' nan '4' '0' '5' 'Unknown' '2' '1'] 


Food and drink
['5' nan '1' '3' 'Great' '4' '2' '0'] 


Inflight entertainment
[ 1. nan  2.  3.  4.  5. 10.  0.] 


Baggage handling
['1' nan '2' '3' '5' '4' 'A'] 


Cleanliness
[ 3. nan  4.  2.  5.  0.] 


Departure Delay in Minutes
359 unique values 


Arrival Delay in Minutes
362 unique values 


Satisfaction
['Dissatisfied' 'neutral' 'satisfied' 'dissatisfied' nan] 




#### 4. The number of missing values of each column.

In [9]:
df_Business.isnull().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

#### 5. Summary statistics for each data column

In [10]:
df_Business.describe() # for numeric columns

Age  Flight Distance  Inflight wifi service  \
count  47373.000000     4.660400e+04           48239.000000   
mean      41.853186     1.934737e+03               2.774519   
std       13.730683     4.369344e+04               1.423917   
min        0.000000    -5.421000e+03               0.000000   
25%       34.000000     5.940000e+02               2.000000   
50%       43.000000     1.597000e+03               3.000000   
75%       52.000000     2.565000e+03               4.000000   
max      185.000000     9.000000e+06               5.000000   

       Departure/Arrival time convenient  Inflight entertainment  \
count                       48239.000000            43649.000000   
mean                            2.910695                3.835391   
std                             1.501711                1.233508   
min                             0.000000                0.000000   
25%                             2.000000                3.000000   
50%                             3.000000                4.000000   
75%                             4.000000                5.000000   
max                             5.000000               10.000000   

        Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  
count  43649.000000                43649.000000              42098.000000  
mean       3.737520                   13.595363                 13.059647  
std        0.987664                   36.595048                 38.116530  
min        0.000000                   -2.000000                -45.000000  
25%        3.000000                    0.000000                  0.000000  
50%        4.000000                    0.000000                  0.000000  
75%        5.000000                   10.000000                 11.000000  
max        5.000000                 1305.000000               1280.000000

In [11]:
df_Business.describe(include = object)

Gender   Customer Type   Type of Travel     Class  \
count    48239           48239            48239     50000   
unique       4               4                3         1   
top     Female  Loyal Customer  Business travel  Business   
freq     23001           40770            44063     50000   

       Ease of Online booking Food and drink Baggage handling Satisfaction  
count                   42705          42705            43649        47840  
unique                      7              7                6            4  
top                         3              4                4    satisfied  
freq                     8635          10890            17056        30400

In [12]:
df_Business.describe(include='all')

Gender   Customer Type           Age   Type of Travel     Class  \
count    48239           48239  47373.000000            48239     50000   
unique       4               4           NaN                3         1   
top     Female  Loyal Customer           NaN  Business travel  Business   
freq     23001           40770           NaN            44063     50000   
mean       NaN             NaN     41.853186              NaN       NaN   
std        NaN             NaN     13.730683              NaN       NaN   
min        NaN             NaN      0.000000              NaN       NaN   
25%        NaN             NaN     34.000000              NaN       NaN   
50%        NaN             NaN     43.000000              NaN       NaN   
75%        NaN             NaN     52.000000              NaN       NaN   
max        NaN             NaN    185.000000              NaN       NaN   

        Flight Distance  Inflight wifi service  \
count      4.660400e+04           48239.000000   
unique              NaN                    NaN   
top                 NaN                    NaN   
freq                NaN                    NaN   
mean       1.934737e+03               2.774519   
std        4.369344e+04               1.423917   
min       -5.421000e+03               0.000000   
25%        5.940000e+02               2.000000   
50%        1.597000e+03               3.000000   
75%        2.565000e+03               4.000000   
max        9.000000e+06               5.000000   

        Departure/Arrival time convenient Ease of Online booking  \
count                        48239.000000                  42705   
unique                                NaN                      7   
top                                   NaN                      3   
freq                                  NaN                   8635   
mean                             2.910695                    NaN   
std                              1.501711                    NaN   
min                              0.000000                    NaN   
25%                              2.000000                    NaN   
50%                              3.000000                    NaN   
75%                              4.000000                    NaN   
max                              5.000000                    NaN   

       Food and drink  Inflight entertainment Baggage handling   Cleanliness  \
count           42705            43649.000000            43649  43649.000000   
unique              7                     NaN                6           NaN   
top                 4                     NaN                4           NaN   
freq            10890                     NaN            17056           NaN   
mean              NaN                3.835391              NaN      3.737520   
std               NaN                1.233508              NaN      0.987664   
min               NaN                0.000000              NaN      0.000000   
25%               NaN                3.000000              NaN      3.000000   
50%               NaN                4.000000              NaN      4.000000   
75%               NaN                5.000000              NaN      5.000000   
max               NaN               10.000000              NaN      5.000000   

        Departure Delay in Minutes  Arrival Delay in Minutes Satisfaction  
count                 43649.000000              42098.000000        47840  
unique                         NaN                       NaN            4  
top                            NaN                       NaN    satisfied  
freq                           NaN                       NaN        30400  
mean                     13.595363                 13.059647          NaN  
std                      36.595048                 38.116530          NaN  
min                      -2.000000                -45.000000          NaN  
25%                       0.000000                  0.000000          NaN  
50%                       0.000000          

#### 6. Frequency distribution (list the number of times each unique value appears) for each data column.

In [13]:
cols = df_Business.columns
for col in cols:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

Gender
Female    23001
Male      22830
F          1361
M          1047
Name: Gender, dtype: int64 

Customer Type
Loyal Customer       40770
Disloyal Customer     6636
disloyal               831
Loyal                    2
Name: Customer Type, dtype: int64 

Age
39.0     1935
40.0     1677
41.0     1533
42.0     1514
44.0     1506
         ... 
73.0       22
78.0       17
85.0       14
180.0       1
185.0       1
Name: Age, Length: 74, dtype: int64 

Type of Travel
Business travel    44063
Business            2122
Personal Travel     2054
Name: Type of Travel, dtype: int64 

Class
Business    50000
Name: Class, dtype: int64 

Flight Distance
337.0     225
2475.0    187
2586.0    148
404.0     133
1744.0    133
         ... 
1386.0      1
938.0       1
2780.0      1
428.0       1
768.0       1
Name: Flight Distance, Length: 3573, dtype: int64 

Inflight wifi service
2.0    10784
3.0    10593
4.0     9140
1.0     8808
5.0     7025
0.0     1889
Name: Inflight wifi service, dtype: int64 

D

#### 7. The number of fully duplicated data rows.

In [14]:
df_Business.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
49995    False
49996    False
49997    False
49998    False
49999    False
Length: 50000, dtype: bool

In [15]:
# Selecting duplicate based on all columns
duplicate = df_Business[df_Business.duplicated()]
  
print("Duplicate Rows :")
duplicate

Duplicate Rows :


Gender      Customer Type   Age   Type of Travel     Class  \
38          F  Disloyal Customer  25.0         Business  Business   
80        NaN                NaN   NaN              NaN  Business   
85        NaN                NaN   NaN              NaN  Business   
168       NaN                NaN   NaN              NaN  Business   
269    Female     Loyal Customer  49.0  Business travel  Business   
...       ...                ...   ...              ...       ...   
49856  Female     Loyal Customer  37.0  Business travel  Business   
49887     NaN                NaN   NaN              NaN  Business   
49908    Male     Loyal Customer  24.0         Business  Business   
49944       M     Loyal Customer  30.0  Business travel  Business   
49959       M     Loyal Customer  30.0  Business travel  Business   

       Flight Distance  Inflight wifi service  \
38                67.0                    1.0   
80                 NaN                    NaN   
85                 NaN                    NaN   
168                NaN                    NaN   
269              101.0                    1.0   
...                ...                    ...   
49856           3987.0                    3.0   
49887              NaN                    NaN   
49908           3993.0                    2.0   
49944              NaN                    3.0   
49959              NaN                    4.0   

       Departure/Arrival time convenient Ease of Online booking  \
38                                   0.0                      1   
80                                   NaN                    NaN   
85                                   NaN                    NaN   
168                                  NaN                    NaN   
269                                  4.0                      4   
...                                  ...                    ...   
49856                                1.0                      3   
49887                                NaN                    NaN   
49908                                5.0                      5   
49944                                3.0                    NaN   
49959                                4.0                    NaN   

      Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
38                 3                     3.0                4          3.0   
80               NaN                     NaN              NaN          NaN   
85               NaN                     NaN              NaN          NaN   
168              NaN                     NaN              NaN          NaN   
269                3                     2.0                1          3.0   
...              ...                     ...              ...          ...   
49856              5                     5.0                5          5.0   
49887            NaN                     NaN              NaN          NaN   
49908              2                     2.0                A          2.0   
49944            NaN                     4.0                1          4.0   
49959            NaN                     4.0                4          4.0   

       Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
38                            0.0                       0.0           NaN  
80                            NaN                       NaN  dissatisfied  
85                            NaN                       NaN  dissatisfied  
168                           NaN                       NaN  dissatisfied  
269                          11.0                      14.0  Dissatisfied  
...                           ...                       ...           ...  
49856                        10.0                      23.0     satisfied  
49887                         NaN                       NaN  dissatisfied  
49908                        23.0                     -45.0  dissatisfied  
49944                        10.0                       NaN     satisfied  
49959      

#### There are 3064 duplicate entries in our dataset

---

## Summary 
### List of the issues that the data contains

Understaning our dataset:
1. In intial dataset, there were 104910 rows and 16 columns in our dataset.

2. We created different dataframe 'df_Business' which contain **50,000 rows and 16 columns**.

3. These columns contain two types of data, which are respectively float and object values.

Data type error:
1. The **Age** should be an integer, but it is in float format. Besides, all rating columns should ideally be integer, but some values such as **Inflight wifi service, Departure/Arrival time convenient, Inflight entertainment, and Cleanliness**, are in float format. 

2. Just as same above, the other rating values such as **Ease of Online booking, Food and drink, and Baggage handling**, should have been in int format, but now they are shown as object format in the dataset because there are some strings such as "Unknown", "Great", and "A" mixing with the rating number. We need to eliminate or change the value of these strings in order to make their value follow the rank from 1 to 5. 

Bad data error: 
1. We also observed unquie values in each column. Except the rating column, **Inflight Wi-Fi Service**, ranging from 0 to 5, all the other rating columns such as **Departure/Arrival Time Convenient, Ease of Online Booking, Food and Drink, Inflight Entertainment, Baggage Handling, and Cleanliness** should not contain 0 in their dataset. 

2. Column Class has a single value, **Business**, which is not giving us more information. Hence, we will eliminate it. 

3. From the frequnecy distribution and statistics, there are some outliers in the column **Age, Flight Distance, Departure Delay in Minutes, and Arrival Delay in Minutes** , thus, we need to depend on unique table and the IQR method to detect these values, and remove or change their value. 

Format error: 
- There is format error in column vlaues of **Gender, Customer Type, Types of Travel, and Satisfaction**. We need to adjust the format of value so as to make sure them have same capital case. 

Missing value: 
- There are missing vlaues in **15 columns**, which conclude all columns except Class column.

Duplicate error: 
- We also have **3064 duplicate rows**. 

---
## 2. Conduct Data Cleansing:
Implement the following tasks:
- Drop the Class column from the created Business Class passenger satisfaction survey (“df_Business”).
- Rename “df_Business” columns by making all the columns’ names in lowercase, and if the column name has whitespace, replace it with an underscore. (For example: “Customer Type” will be renamed to “customer_type”).
- Modify improperly formatted data and handle bad data.
- Handle the missing values (You should have zero missing values in your cleaned dataset).
- Remove the fully duplicated observations.
- Fix the columns datatypes.

---
### 1. Drop the Class column from the created Business Class passenger satisfaction survey (“df_Business”).

In [16]:
# Dropping columns
df_Business.drop([ 'Class'], axis=1, inplace=True)
df_Business = df_Business.reset_index(drop=True)  # To reset the df_sales index
df_Business

/Users/mac/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Gender      Customer Type   Age   Type of Travel  Flight Distance  \
0        Male     Loyal Customer  69.0  Personal Travel          -5421.0   
1        Male  Disloyal Customer  41.0  Business travel              0.0   
2        Male     Loyal Customer  52.0  Business travel              0.0   
3      Female  Disloyal Customer  21.0  Business travel              0.0   
4      Female     Loyal Customer  57.0  Business travel              0.0   
...       ...                ...   ...              ...              ...   
49995    Male     Loyal Customer  45.0  Business travel           3305.0   
49996  Female     Loyal Customer  35.0  Business travel            337.0   
49997    Male     Loyal Customer  47.0  Business travel           3496.0   
49998    Male     Loyal Customer  42.0  Business travel           3605.0   
49999  Female           disloyal   0.0  Business travel            862.0   

       Inflight wifi service  Departure/Arrival time convenient  \
0                        3.0                                2.0   
1                        1.0                                1.0   
2                        2.0                                4.0   
3                        4.0                                0.0   
4                        3.0                                4.0   
...                      ...                                ...   
49995                    2.0                                2.0   
49996                    5.0                                5.0   
49997                    2.0                                2.0   
49998                    4.0                                4.0   
49999                    5.0                                5.0   

      Ease of Online booking Food and drink  Inflight entertainment  \
0                          3              5                     1.0   
1                        NaN            NaN                     NaN   
2                          4              1                     2.0   
3                        NaN            NaN                     NaN   
4                          4              5                     2.0   
...                      ...            ...                     ...   
49995                      2              3                     4.0   
49996                      5              2                     4.0   
49997                      2              2                     5.0   
49998                      4              3                     4.0   
49999                      5              5                     5.0   

      Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                    1          3.0                        14.0   
1                  NaN          NaN                         NaN   
2                    2          4.0                         0.0   
3                  NaN          NaN                         NaN   
4                    3          2.0                         0.0   
...                ...          ...                         ...   
49995                4          3.0                         8.0   
49996                4          5.0                        54.0   
49997                5          4.0                         0.0   
49998                4          3.0                         0.0   
49999                1          5.0                         9.0   

       Arrival Delay in Minutes  Satisfaction  
0                           0.0  Dissatisfied  
1                           NaN       neutral  
2                           6.0  Dissatisfied  
3                           NaN       neutral  
4                           0.0  Dissatisfied  
...                         ...           ...  
49995                       0.0     satisfied  
49996                      40.0     satisfied  
49997                       2.0     satisfied  
49998                       0.0     satisfied  
49999                       5.0       neutral  

[50000 rows x 15 columns]

---
### 2. Rename “df_Business” columns by making all the columns’ names in lowercase, and if the column name has whitespace, replace it with an underscore. 

In [17]:
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Flight Distance',
       'Inflight wifi service', 'Departure/Arrival time convenient',
       'Ease of Online booking', 'Food and drink', 'Inflight entertainment',
       'Baggage handling', 'Cleanliness', 'Departure Delay in Minutes',
       'Arrival Delay in Minutes', 'Satisfaction'],
      dtype='object')

In [18]:
df_Business.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)  
df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'flight_distance',
       'inflight_wifi_service', 'departure/arrival_time_convenient',
       'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment',
       'baggage_handling', 'cleanliness', 'departure_delay_in_minutes',
       'arrival_delay_in_minutes', 'satisfaction'],
      dtype='object')

In [19]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             48239 non-null  object 
 1   customer_type                      48239 non-null  object 
 2   age                                47373 non-null  float64
 3   type_of_travel                     48239 non-null  object 
 4   flight_distance                    46604 non-null  float64
 5   inflight_wifi_service              48239 non-null  float64
 6   departure/arrival_time_convenient  48239 non-null  float64
 7   ease_of_online_booking             42705 non-null  object 
 8   food_and_drink                     42705 non-null  object 
 9   inflight_entertainment             43649 non-null  float64
 10  baggage_handling                   43649 non-null  object 
 11  cleanliness                        43649 non-null  flo

---
### 3. Modify improperly formatted data and handle bad data.
- Fixing the data format below:
   - Gender
   - Customer Type
   - Types of Travel
   - Satisfaction

In [20]:
# Printing unique values for list of columns
for col in ['gender', 'customer_type', 'type_of_travel', 'satisfaction']:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')

gender
#.of unique values:  5 
List of unique values:  ['Male' 'Female' 'F' 'M' nan] 
*************************************

customer_type
#.of unique values:  5 
List of unique values:  ['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
*************************************

type_of_travel
#.of unique values:  4 
List of unique values:  ['Personal Travel' 'Business travel' 'Business' nan] 
*************************************

satisfaction
#.of unique values:  5 
List of unique values:  ['Dissatisfied' 'neutral' 'satisfied' 'dissatisfied' nan] 
*************************************



In [21]:
# Replacing values
df_Business['gender'] = df_Business['gender'].replace(['F', 'M'],['Female', 'Male'])
df_Business['customer_type'] = df_Business['customer_type'].replace(['Loyal','disloyal'],['Loyal Customer','Disloyal Customer'])
df_Business['type_of_travel'] = df_Business['type_of_travel'].replace(['Business','Business travel' ],['Business Travel', 'Business Travel'])
df_Business['satisfaction'] = df_Business['satisfaction'].replace(['dissatisfied', 'satisfied','neutral'],['Dissatisfied', 'Satisfied','Neutral'])

In [22]:
# Printing unique values for list of columns
# for each column
for col in ['gender', 'customer_type', 'type_of_travel', 'satisfaction']:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')


gender
#.of unique values:  3 
List of unique values:  ['Male' 'Female' nan] 
*************************************

customer_type
#.of unique values:  3 
List of unique values:  ['Loyal Customer' 'Disloyal Customer' nan] 
*************************************

type_of_travel
#.of unique values:  3 
List of unique values:  ['Personal Travel' 'Business Travel' nan] 
*************************************

satisfaction
#.of unique values:  4 
List of unique values:  ['Dissatisfied' 'Neutral' 'Satisfied' nan] 
*************************************



---
- Fixing the bad data(**Bad Value**) in the columns below
  - **departure/arrival_time_convenient** (Has 0 value)
  - **ease_of_online_booking** (Has Unknown and 0 value)
  - **food_and_drink** (Has Great and 0 value)
  - **inflight_entertainment** (Has 10 and 0 value)
  - **baggage_handling** (Has A value)
  - **cleanliness** (Has 0 value)
- We use mode value of each column above to replace 0 value, and as for "Great", "A", "10", these values all represent positive attitude of customers, so we use 5 to replace them. 

In [23]:
# Printing unique values for list of columns
for col in ['departure/arrival_time_convenient', 'ease_of_online_booking', 'food_and_drink', 
            'inflight_entertainment', 'baggage_handling', 'cleanliness']:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')

departure/arrival_time_convenient
#.of unique values:  7 
List of unique values:  [ 2.  1.  4.  0.  5.  3. nan] 
*************************************

ease_of_online_booking
#.of unique values:  8 
List of unique values:  ['3' nan '4' '0' '5' 'Unknown' '2' '1'] 
*************************************

food_and_drink
#.of unique values:  8 
List of unique values:  ['5' nan '1' '3' 'Great' '4' '2' '0'] 
*************************************

inflight_entertainment
#.of unique values:  8 
List of unique values:  [ 1. nan  2.  3.  4.  5. 10.  0.] 
*************************************

baggage_handling
#.of unique values:  7 
List of unique values:  ['1' nan '2' '3' '5' '4' 'A'] 
*************************************

cleanliness
#.of unique values:  6 
List of unique values:  [ 3. nan  4.  2.  5.  0.] 
*************************************



In [24]:
# Print mode value for each column
for col in ['departure/arrival_time_convenient', 'ease_of_online_booking', 'food_and_drink', 
            'inflight_entertainment', 'baggage_handling', 'cleanliness']:    
    mode_value = df_Business[col].mode()
    print('The mode of ', col, '\nis: ',mode_value[0], '\n*************************************\n')

The mode of  departure/arrival_time_convenient 
is:  4.0 
*************************************

The mode of  ease_of_online_booking 
is:  3 
*************************************

The mode of  food_and_drink 
is:  4 
*************************************

The mode of  inflight_entertainment 
is:  4.0 
*************************************

The mode of  baggage_handling 
is:  4 
*************************************

The mode of  cleanliness 
is:  4.0 
*************************************



In [25]:
# Replacing values
df_Business['departure/arrival_time_convenient'] = df_Business['departure/arrival_time_convenient'].replace([0],[4])
df_Business['ease_of_online_booking'] = df_Business['ease_of_online_booking'].replace(['0'],['3'])
df_Business['food_and_drink'] = df_Business['food_and_drink'].replace(['Great','0' ],['5', '4'])
df_Business['inflight_entertainment'] = df_Business['inflight_entertainment'].replace([10, 0],[5, 4])
df_Business['baggage_handling'] = df_Business['baggage_handling'].replace(['A'],['5'])
df_Business['cleanliness'] = df_Business['cleanliness'].replace([0],[4])

In [26]:
# Printing unique values for list of columns after replace
for col in ['departure/arrival_time_convenient', 'ease_of_online_booking', 'food_and_drink', 
            'inflight_entertainment', 'baggage_handling', 'cleanliness']:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')

departure/arrival_time_convenient
#.of unique values:  6 
List of unique values:  [ 2.  1.  4.  5.  3. nan] 
*************************************

ease_of_online_booking
#.of unique values:  7 
List of unique values:  ['3' nan '4' '5' 'Unknown' '2' '1'] 
*************************************

food_and_drink
#.of unique values:  6 
List of unique values:  ['5' nan '1' '3' '4' '2'] 
*************************************

inflight_entertainment
#.of unique values:  6 
List of unique values:  [ 1. nan  2.  3.  4.  5.] 
*************************************

baggage_handling
#.of unique values:  6 
List of unique values:  ['1' nan '2' '3' '5' '4'] 
*************************************

cleanliness
#.of unique values:  5 
List of unique values:  [ 3. nan  4.  2.  5.] 
*************************************



---
- Fixing the bad data(**Outliers**) in the columns below
 - **age** (Has extrame value and 0)
 - **flight_distance** (Has outliers)
 - **departure_delay_in_minutes** (Has negative value)
 - **arrival_delay_in_minutes** (Has negative value)
- We detect outliers in the numerical column above by **leveraging unique value and IQR approach**. Then, we use mode number of age to replace its outliers, use median number of flight_distance to replace its outliers, and use 0 to replace departure_delay_in_minutes, and arrival_delay_in_minutes' negative value. 

**For age column**

In [27]:
# Check the unque value of age column in order to find some outliers
a = df_Business['age'].unique()
print(sorted(a, reverse=True))
# We find there are three outliers in the age column, which are respectively 185, 180, and 0, because they are counter-intuitive.

[185.0, 69.0, 60.0, 57.0, 56.0, 50.0, nan, 180.0, 85.0, 80.0, 79.0, 78.0, 77.0, 76.0, 75.0, 74.0, 73.0, 72.0, 71.0, 70.0, 68.0, 67.0, 66.0, 65.0, 64.0, 63.0, 62.0, 61.0, 59.0, 58.0, 55.0, 54.0, 53.0, 52.0, 51.0, 49.0, 48.0, 47.0, 46.0, 45.0, 44.0, 43.0, 42.0, 41.0, 40.0, 39.0, 38.0, 37.0, 36.0, 35.0, 34.0, 33.0, 32.0, 30.0, 27.0, 26.0, 25.0, 24.0, 22.0, 21.0, 20.0, 19.0, 18.0, 17.0, 16.0, 15.0, 14.0, 13.0, 12.0, 11.0, 10.0, 9.0, 8.0, 7.0, 0.0]


In [28]:
# Find the mode value of age
mode_value = df_Business['age'].mode()
mode_value[0]

39.0

In [29]:
# replace 0, 180, 185 by 39
df_Business['age'] = df_Business['age'].replace([0.0,185.0, 180.0],[39.0,39.0,39.0])

In [30]:
# Check the unque value of age column again
df_Business['age'].unique()

array([69., 41., 52., 21., 57., 45., 39., 33., 10., 55., 59., 53., 56.,
       47., 54., 25., 51., 60., 27., 43., 20., 19., 63., 50., nan, 38.,
       46., 85., 26., 42., 30., 24., 22., 17., 48., 40., 16., 32., 35.,
       44., 49., 65., 67., 66., 64.,  7., 37., 36., 15., 62., 34., 68.,
       70., 58., 11., 74., 13., 80., 75.,  9., 61., 71., 12., 73., 72.,
       18., 79.,  8., 76., 14., 77., 78.])

**For flight_distance column**

In [31]:
Q1 = df_Business['flight_distance'].quantile(0.25)
Q3 = df_Business['flight_distance'].quantile(0.75)
IQR = Q3 - Q1
lower_limit = Q1 - 1.5*IQR
upper_limit = Q3 + 1.5*IQR
    
df_outlier = df_Business[(df_Business['flight_distance']<lower_limit)|(df_Business['flight_distance']>upper_limit)]
print("The number of outlier of column " + 'flight_distance' + " is" +" "+ str(len(df_outlier)), '\n*************************************\n')
    
print(df_outlier.flight_distance,'\n*************************************\n')


The number of outlier of column flight_distance is 3 
*************************************

0          -5421.0
49993    2820000.0
49994    9000000.0
Name: flight_distance, dtype: float64 
*************************************



In [32]:
# Find the median value of age
median_value = df_Business['flight_distance'].median()
median_value

1597.0

In [33]:
# replace the outliers by 1597
df_Business['flight_distance'] = df_Business['flight_distance'].replace([-5421.0,2820000.0, 9000000.0],[1597.0,1597.0,1597.0])

In [34]:
df_outlier = df_Business[(df_Business['flight_distance']<lower_limit)|(df_Business['flight_distance']>upper_limit)]
print("The number of outlier of column " + 'flight_distance' + " is" +" "+ str(len(df_outlier)), '\n*************************************\n')


The number of outlier of column flight_distance is 0 
*************************************



**For departure_delay_in_minutes and arrival_delay_in_minutes column**

In [35]:
# Check the negative number in these two columns. 
a = df_Business.loc[df_Business.departure_delay_in_minutes < 0]
b = df_Business.loc[df_Business.arrival_delay_in_minutes < 0]
print("The negative value are:", a.departure_delay_in_minutes.unique(), b.arrival_delay_in_minutes.unique())

The negative value are: [-2.] [-45.]


In [36]:
# The negative numbers of above may represent that an airplane arrive or depart in advance, which may raise reader's confusions to understand the info of columns.
# We choose use 0 to replace -2 and -45. 
df_Business['departure_delay_in_minutes'] = df_Business['departure_delay_in_minutes'].replace([-2.0],[0])
df_Business['arrival_delay_in_minutes'] = df_Business['arrival_delay_in_minutes'].replace([-45.0],[0])

In [37]:
# Check the negative number in these two columns. 
a = df_Business.loc[df_Business.departure_delay_in_minutes < 0]
b = df_Business.loc[df_Business.arrival_delay_in_minutes < 0]
print("The negative values are replaced", a.departure_delay_in_minutes.unique(), b.arrival_delay_in_minutes.unique())

The negative values are replaced [] []


---
### 4.Handle the missing values (You should have zero missing values in your cleaned dataset).

#### The Method to deal with missing values in below columns:
- **gender**: we use "Unknown" to replace NA.
- **customer_type**: we use "Unknown" to replace NA.
- **age**: we use mean of age to replace NA.
- **type_of_travel**: we use "Unknown" to replace NA.
- **flight_distance**: we use mode of distance to replace NA.
- **inflight_wifi_service**: we use mode of rank to replace NA.
- **departure/arrival_time_convenient**: we use mode of rank to replace NA.
- **ease_of_online_booking**: we use mode of rank to replace NA.
- **food_and_drink**: we use mode of rank to replace NA.
- **inflight_entertainment**: we use mode of rank to replace NA.
- **baggage_handling**: we use mode of rank to replace NA.
- **cleanliness**: we use mode of rank to replace NA.
- **departure_delay_in_minutes**: we use mean of minutes to replace NA.
- **arrival_delay_in_minutes**: we use mean of minutes to replace NA.
- **satisfaction**: we use "Unknown" tp replace NA. 

In [38]:
# Number of missing values in each coulmn
df_Business.isnull().sum()

gender                               1761
customer_type                        1761
age                                  2627
type_of_travel                       1761
flight_distance                      3396
inflight_wifi_service                1761
departure/arrival_time_convenient    1761
ease_of_online_booking               7295
food_and_drink                       7295
inflight_entertainment               6351
baggage_handling                     6351
cleanliness                          6351
departure_delay_in_minutes           6351
arrival_delay_in_minutes             7902
satisfaction                         2160
dtype: int64

In [39]:
# Replace "Unknown" for missing value in column gender, customer_type, type_of_travel and satisfaction
df_Business['gender'].fillna('Unknown', inplace=True)
df_Business['customer_type'].fillna('Unknown', inplace=True)
df_Business['type_of_travel'].fillna('Unknown', inplace=True)
df_Business['satisfaction'].fillna('Unknown', inplace=True)

In [40]:
# Find the average for a datafram column,age
mean_value = round(df_Business['age'].mean(),0)
print('The mean of age is: ', mean_value)
# Repalcing missing values in age with the mean
df_Business['age'].fillna(mean_value, inplace=True)
df_Business['age'].unique()

The mean of age is:  43.0


array([69., 41., 52., 21., 57., 45., 39., 33., 10., 55., 59., 53., 56.,
       47., 54., 25., 51., 60., 27., 43., 20., 19., 63., 50., 38., 46.,
       85., 26., 42., 30., 24., 22., 17., 48., 40., 16., 32., 35., 44.,
       49., 65., 67., 66., 64.,  7., 37., 36., 15., 62., 34., 68., 70.,
       58., 11., 74., 13., 80., 75.,  9., 61., 71., 12., 73., 72., 18.,
       79.,  8., 76., 14., 77., 78.])

In [41]:
# Replace mode for missing value in column below columns.
for col in ['flight_distance', 'inflight_wifi_service', 'departure/arrival_time_convenient', 
            'ease_of_online_booking','food_and_drink','inflight_entertainment', 'baggage_handling', 'cleanliness']: 
    mode_value = df_Business[col].mode()
    
    # Repalcing missing values in ranking columns with the mode
    df_Business[col].fillna(mode_value[0], inplace=True)

In [42]:
# Replace mean for missing value in column below columns.
for col in ['departure_delay_in_minutes', 'arrival_delay_in_minutes']: 
    mean_value = round(df_Business[col].mean(),0)
    df_Business[col].fillna(mean_value, inplace=True)

In [43]:
# Number of missing values in each coulmn after cleaning
df_Business.isnull().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
baggage_handling                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64

---
### 5. Remove the fully duplicated observations.

In [44]:
# Print the number of duplicate observations.
df_Business.duplicated().sum()

3071

In [45]:
#Drop Duplicate Observations.
df_Business.drop_duplicates(inplace=True)

In [46]:
print("Re-checking the number of duplicate observations:")
df_Business.duplicated().sum()

Re-checking the number of duplicate observations:


0

In [47]:
df_Business.shape

(46929, 15)

---
### 6. Fix the columns datatypes.

In [48]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46929 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             46929 non-null  object 
 1   customer_type                      46929 non-null  object 
 2   age                                46929 non-null  float64
 3   type_of_travel                     46929 non-null  object 
 4   flight_distance                    46929 non-null  float64
 5   inflight_wifi_service              46929 non-null  float64
 6   departure/arrival_time_convenient  46929 non-null  float64
 7   ease_of_online_booking             46929 non-null  object 
 8   food_and_drink                     46929 non-null  object 
 9   inflight_entertainment             46929 non-null  float64
 10  baggage_handling                   46929 non-null  object 
 11  cleanliness                        46929 non-null  flo

By going through data type above, we need to fix all the float type into integer type. For the columns 'ease_of_online_booking', 'food_and_drink', 'baggage_handling', we need to pay more attention on these data types and figure out what's wrong with their data types. 

In [49]:
for col in ['ease_of_online_booking', 'food_and_drink', 'baggage_handling']:
    print(col)
    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')

ease_of_online_booking
#.of unique values:  6 
List of unique values:  ['3' '4' '5' 'Unknown' '2' '1'] 
*************************************

food_and_drink
#.of unique values:  5 
List of unique values:  ['5' '4' '1' '3' '2'] 
*************************************

baggage_handling
#.of unique values:  5 
List of unique values:  ['1' '4' '2' '3' '5'] 
*************************************



We find ease_of_online_booking is composed with number and strings, so we have to change the "Unknown" into number for making the final data type become integer. 

In [50]:
mode_value = df_Business['ease_of_online_booking'].mode()
mode_value[0]

'3'

In [51]:
df_Business['ease_of_online_booking'] = df_Business['ease_of_online_booking'].replace(['Unknown' ],['3'])

Convert string value into numeric value

In [52]:
df_Business['ease_of_online_booking'] = pd.to_numeric(df_Business.ease_of_online_booking)
df_Business['food_and_drink'] = pd.to_numeric(df_Business.food_and_drink)
df_Business['baggage_handling'] = pd.to_numeric(df_Business.baggage_handling)

Convert all float value into integer value

In [53]:
df_Business = df_Business.convert_dtypes()
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46929 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   gender                             46929 non-null  string
 1   customer_type                      46929 non-null  string
 2   age                                46929 non-null  Int64 
 3   type_of_travel                     46929 non-null  string
 4   flight_distance                    46929 non-null  Int64 
 5   inflight_wifi_service              46929 non-null  Int64 
 6   departure/arrival_time_convenient  46929 non-null  Int64 
 7   ease_of_online_booking             46929 non-null  Int64 
 8   food_and_drink                     46929 non-null  Int64 
 9   inflight_entertainment             46929 non-null  Int64 
 10  baggage_handling                   46929 non-null  Int64 
 11  cleanliness                        46929 non-null  Int64 
 12  depa

--- 
## 3. Export the cleaned dataset to CSV file

In [54]:
df_Business.reset_index(drop=True, inplace = True)
df_Business

gender      customer_type  age   type_of_travel  flight_distance  \
0        Male     Loyal Customer   69  Personal Travel             1597   
1        Male  Disloyal Customer   41  Business Travel                0   
2        Male     Loyal Customer   52  Business Travel                0   
3      Female  Disloyal Customer   21  Business Travel                0   
4      Female     Loyal Customer   57  Business Travel                0   
...       ...                ...  ...              ...              ...   
46924    Male     Loyal Customer   45  Business Travel             3305   
46925  Female     Loyal Customer   35  Business Travel              337   
46926    Male     Loyal Customer   47  Business Travel             3496   
46927    Male     Loyal Customer   42  Business Travel             3605   
46928  Female  Disloyal Customer   39  Business Travel              862   

       inflight_wifi_service  departure/arrival_time_convenient  \
0                          3                                  2   
1                          1                                  1   
2                          2                                  4   
3                          4                                  4   
4                          3                                  4   
...                      ...                                ...   
46924                      2                                  2   
46925                      5                                  5   
46926                      2                                  2   
46927                      4                                  4   
46928                      5                                  5   

       ease_of_online_booking  food_and_drink  inflight_entertainment  \
0                           3               5                       1   
1                           3               4                       4   
2                           4               1                       2   
3                           3               4                       4   
4                           4               5                       2   
...                       ...             ...                     ...   
46924                       2               3                       4   
46925                       5               2                       4   
46926                       2               2                       5   
46927                       4               3                       4   
46928                       5               5                       5   

       baggage_handling  cleanliness  departure_delay_in_minutes  \
0                     1            3                          14   
1                     4            4                          14   
2                     2            4                           0   
3                     4            4                          14   
4                     3            2                           0   
...                 ...          ...                         ...   
46924                 4            3                           8   
46925                 4            5                          54   
46926                 5            4                           0   
46927                 4            3                           0   
46928                 1            5                           9   

       arrival_delay_in_minutes  satisfaction  
0                             0  Dissatisfied  
1                            14       Neutral  
2                             6  Dissatisfied  
3                            14       Neutral  
4                             0  Dissatisfied  
...                         ...           ...  
46924                         0     Satisfied  
46925                        40     Satisfied  
46926                         2     Satisfied  
46927                         0     Satisfied  
46928                         5       Neutral  

[46929 rows x 15 columns]

In [55]:
#Exporting data into csv file
df_Business.to_csv("Project_2_Group15.csv", index= False) 

--- 
## 4. Compare the Data Characteristics Before and After Data Cleansing

### This is the data frame(df_Business) after data cleansing. 
##### You will observe these changes:

- **Data Frame**:
    - Total data points were reduced from 104,910 to 50,000 as we filterd the data based on "Class" == "Business".
    - Initially there were 16 columns of which we dropped one unnecessary columns 'Class'. This column had only single value from which we were not getting any usefull information. Hence, we now have 15 columns in our dataframe.
    - We also observed unquie values and count for each column. 
    - We made all the columns’ names in lowercase, and for those the column name with whitespace, we replaced it with an underscore. 
    - Our final data frame is composed of 15 columns and 46929 rows. We try our best to keep as more data as possible while ensuring data integrity, accuracy, and representation.
- **Format Adjustment**:
    - Merged similar gender into one. Ex: "Female" & "F", "Male"&"M". This was part formating imporperly formatted data in column to avoid unnecessary confusion.
    - Change similiar values into one unified form. Ex: 'Business travel' &'Business'. It can help us to reduce the redundent values. 

- **Datatype Change**:
    - We changed all ranking columns from object/float into int. This is because these columns are representing rating which should be idealy integer value. 
    - We also changed the data type of age, flight distance, departure delay and arrive delay from float into int, because these columns didn't contain any of float numbers. 
    
- **Bad Data**:
    - we fixed the bad data(Bad Value) in the columns below. We use mode value of each column above to replace 0 value, and as for "Great", "A", "10", these values all represent positive attitude of customers, so we use 5 to replace them.
        - departure/arrival_time_convenient (Has 0 value),
        - ease_of_online_booking (Has Unknown and 0 value), 
        - food_and_drink (Has Great and 0 value), 
        - inflight_entertainment (Has 10 and 0 value),
        - baggage_handling (Has A value),
        - cleanliness (Has 0 value). 
    - we fixed the bad data(Outliers) in the columns. We detect outliers in the numerical column below by leveraging unique value and IQR approach. Then, we use mode number of age to replace its outliers, use median number of flight_distance to replace its outliers, and use 0 to replace departure_delay_in_minutes, and arrival_delay_in_minutes' negative value.
        - age (Has extrame value and 0), 
        - flight_distance (Has outliers), 
        - departure_delay_in_minutes (Has negative value), 
        - arrival_delay_in_minutes (Has negative value).

- **Missing Value**:
    - Replaced NA values with appropriate statistics. Null values in column below were replaced with "Unknown".
        - gender 
        - customer_type
        - type_of_travel 
        - satisfaction 
    
    - Missing value in columns below were replaced with mode value. This decision was taken as these columns represent rating or distance.
        - 'flight_distance'
        - 'inflight_wifi_service'
        - 'departure/arrival_time_convenient'
        - 'ease_of_online_booking'
        - 'food_and_drink'
        - 'inflight_entertainment'
        - 'baggage_handling'
        - 'cleanliness'  
    - Missing value in columns below were replaced with mean value. This was because after dealing with bad values of these column, mean could provid a better central tendency of these values. 
        - 'age'
        - 'departure_delay_in_minutes'
        - 'arrival_delay_in_minutes' 
    
- **Duplicate Rows**:
    - Initially, there were 3064 duplicate rows. However, after series of cleaning processes such as replacing NA values or change bad values, there were new 7 duplicate rows appearing in the data frame. Finally, there were 3071 duplicate rows which were later removed. 


## * Before Cleansing *

### Columns names and datatype


|  Column                   | Dtype  
|  :------                  |   -----:
|  Gender                   |   object  
|  Customer Type            |   object 
|  Age                      |   float64 
|  Type of Travel           |   object 
|  Class                    |   object 
|  Flight Distance          |   float64
|  Inflight wifi service    |   float64 
|  Departure/Arrival time convenient |   float64  
|  Ease of Online booking   |   object 
|  Food and drink           |   object  
|  Inflight entertainment   |   float64 
|  Baggage handling         |   object  
|  Cleanliness              |   float64 
|  Departure Delay in Minutes|  float64 
|  Arrival Delay in Minutes |   float64 
|  Satisfaction             |   object  

### Unique Values for each Column



|  Column                   | Unique Value  
|  :------                  |   -----:
|  Gender                   |   'Male' 'Female' 'F' 'M' nan  
|  Customer Type            |   'Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal' 
|  Age                      |   75 unique values 
|  Type of Travel           |   'Personal Travel' 'Business travel' 'Business' nan 
|  Class                    |   'Business' 
|  Flight Distance          |   3574 unique values
|  Inflight wifi service    |   3.  1.  2.  4.  0.  5. nan 
|  Departure/Arrival time convenient |   2.  1.  4.  0.  5.  3. nan  
|  Ease of Online booking   |   '3' nan '4' '0' '5' 'Unknown' '2' '1' 
|  Food and drink           |   '5' nan '1' '3' 'Great' '4' '2' '0'  
|  Inflight entertainment   |    1. nan  2.  3.  4.  5. 10.  0. 
|  Baggage handling         |   '1' nan '2' '3' '5' '4' 'A'  
|  Cleanliness              |   3. nan  4.  2.  5.  0. 
|  Departure Delay in Minutes|  359 unique values 
|  Arrival Delay in Minutes |   362 unique values
|  Satisfaction             |   'Dissatisfied' 'neutral' 'satisfied' 'dissatisfied' nan

### Number of Missing Values for each Column


|  Column                   | Missing Value  
|  :------                  |   -----:
|  Gender                   |   1761  
|  Customer Type            |   1761
|  Age                      |   2627 
|  Type of Travel           |   1761
|  Class                    |   0 
|  Flight Distance          |   3396
|  Inflight wifi service    |   1761
|  Departure/Arrival time convenient |   1761 
|  Ease of Online booking   |   7295 
|  Food and drink           |   7295  
|  Inflight entertainment   |   6351 
|  Baggage handling         |   6351  
|  Cleanliness              |   6351 
|  Departure Delay in Minutes|  6351 
|  Arrival Delay in Minutes |   7902 
|  Satisfaction             |   2160  


---
## *** After Cleansing ***

### Columns names and datatype


|  Column                   | Dtype  
|  :------                  |   -----:
|  Gender                   |   string  
|  Customer Type            |   string 
|  Age                      |   Int64 
|  Type of Travel           |   string 
|  Flight Distance          |   Int64
|  Inflight wifi service    |   Int64 
|  Departure/Arrival time convenient |   Int64  
|  Ease of Online booking   |   Int64 
|  Food and drink           |   Int64  
|  Inflight entertainment   |   Int64 
|  Baggage handling         |   Int64  
|  Cleanliness              |   Int64 
|  Departure Delay in Minutes|  Int64 
|  Arrival Delay in Minutes |   Int64 
|  Satisfaction             |   string  


### Unique Values for each Column


|  Column                   | Unique Value  
|  :------                  |   -----:
|  Gender                   |   'Male', 'Female', 'Unknown' 
|  Customer Type            |   'Loyal Customer', 'Disloyal Customer', 'Unknown'
|  Age                      |   71 unique values 
|  Type of Travel           |   'Personal Travel', 'Business Travel', 'Unknown'
|  Flight Distance          |   3570 unique values
|  Inflight wifi service    |   3, 1, 2, 4, 0, 5
|  Departure/Arrival time convenient |  2, 1, 4, 5, 3 
|  Ease of Online booking   |   3, 4, 5, 2, 1
|  Food and drink           |   5, 4, 1, 3, 2 
|  Inflight entertainment   |   1, 4, 2, 3, 5 
|  Baggage handling         |   1, 4, 2, 3, 5  
|  Cleanliness              |   3, 4, 2, 5
|  Departure Delay in Minutes|  357 unique values 
|  Arrival Delay in Minutes |   360 unique values
|  Satisfaction             |   'Dissatisfied', 'Neutral', 'Satisfied', 'Unknown'


### Number of Missing Values for each Column


|  Column                   | Missing Value  
|  :------                  |   -----:
|  Gender                   |   0  
|  Customer Type            |   0
|  Age                      |   0 
|  Type of Travel           |   0
|  Class                    |   0 
|  Flight Distance          |   0
|  Inflight wifi service    |   0
|  Departure/Arrival time convenient |   0
|  Ease of Online booking   |   0 
|  Food and drink           |   0  
|  Inflight entertainment   |   0 
|  Baggage handling         |   0  
|  Cleanliness              |   0 
|  Departure Delay in Minutes|  0 
|  Arrival Delay in Minutes |   0 
|  Satisfaction             |   0  